### PR TITLE
return_scaled - bindat, neuroscope, and intan

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
@alejoe91 Got around to updating this per our discussion

### CODE REVIEW

File: spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py [hunk 1]
```diff
@@ -61,16 +63,29 @@ def get_sampling_frequency(self):
         return self._fs
 
     @check_get_traces_args
-    def get_traces(self, channel_ids=None, start_frame=None, end_frame=None, return_scaled=True, dtype=None):
+    def get_traces(self, channel_ids=None, start_frame=None, end_frame=None, return_scaled: Optional[bool] = None,
```
Comment (alejoe91) [inline spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py:66]: @CodyCBakerPhD can you make rerurn_scaled not optional and set it to True by default?

Comment (CodyCBakerPhD) [reply spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py:66]: @alejoe91 I played around with that at first but it makes it more complicated to error check without being able to tell if the user manually specified it as True or False, or if it is in default mode. Also note that it is always 'optional' throughout the repo, since it has a default behavior (True). That behavior is mirrored here, and everything in the default mode acts as though it were True. In order to provide the most robust checking against the potentially conflicting `dtype` argument, it really helps to be able to tell when the user sets it, which we can't tell if we just leave it as `True` and allow `dtype` to pass along with an unspecified `return_scale`.

For example, a user that wanted to try to get the base `int` type returned might call

```
my_intan_re = se.IntanRecordingExtractor(intan_file_path)
my_intan_re.get_traces(dtype='int16')
```

but if `return_scaled=True` by default, it would either have to get overridden by the `dtype` (which might imply that the result is scaled, but also somehow an `int` type), or return an error telling the user they have to manually specify both `return_scaled` and `dtype` at `get_traces`. Likewise they could do

```
my_intan_re = se.IntanRecordingExtractor(intan_file_path, dtype='int16')
my_intan_re.get_traces() or my_intan_re.get_traces(return_scaled=True)
```

which, if `return_scaled=True` by default, would force adherence to that and return float traces, which could be quite confusing.

Basically, the way it works right now is, by detecting which arguments were actually specified by the user, we can both provide a clear path of which arguments override each other based on what the user is trying to do (i.e., if `return_scaled` is specified but `dtype` is not, use the implied behavior of `return_scaled`; if `dtype` is specified but `return_scaled` is not, use the `dtype` behavior). If neither is specified in `get_traces` we use the internal `dtype` set at initialization. If both are passed, and conflict with each other, an informative message is given in this respect.

Comment (alejoe91) [reply spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py:66]: Thanks @CodyCBakerPhD, but in this case `return_scaled: Optional[bool] = None` means that if the user doesn't pass it, it is `None`. IMO, in order to get `int16`, the user should do:
```
my_intan_re.get_traces(return_scaled=False, dtype='int16')
```

Since `return_scaled` should be `True` by default, I think that:
```
my_intan_re.get_traces(dtype='int16')
```

Should throw an error. In case `return_scaled` is `False` but `dtype` is not set, I would return `uint16`, which is how the data is stored. 

Also, since we have all this options for `get_traces`, I think we can remove `dtype` as an argument to the `__init__`. What do you think?

File: spikeextractors/extractors/bindatrecordingextractor/bindatrecordingextractor.py [hunk 2]
```diff
@@ -42,12 +42,17 @@ def __init__(self, file_path, sampling_frequency, numchan, dtype, recording_chan
...
+
         if 'numpy' in str(dtype):
             dtype_str = str(dtype).replace("<class '", "").replace("'>", "")
             # drop 'numpy
             dtype_str = dtype_str.split('.')[1]
         else:
             dtype_str = str(dtype)
+
+        if gain is not None:
+            self.set_channel_gains(channel_ids=self.get_channel_ids(), gains=gain)
```
Comment (alejoe91) [inline spikeextractors/extractors/bindatrecordingextractor/bindatrecordingextractor.py:54]: good, but we should also update the `get_traces` to use the `gain` property instead of the internal `_gain` field

File: spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py:66]: > Thanks @CodyCBakerPhD, but in this case return_scaled: Optional[bool] = None means that if the user doesn't pass it, it is None.

Yes - but if you read the code below, if it is `None` there are a couple ways it then defaults based on if `dtype` was specified. If you track the default mode from top-down (including `__init__`), it delivers exactly the same as every other extractor as if `return_scaled` was `True`, but we now have that extra piece of information that it was actually passed by the user and not simply defaulted.

While I'm not against removing `dtype` from `__init__` since it really doesn't do much, just keep in mind that removing that breaks back-compatibility, which I'm really trying to preserve here. Also, if we did that, the default `dtype` at `__init__` would have to be `uint16` in order to preserve the gain and offset information.

Another break in back-compatibility also occurs with what you propose on forcing the pairing of `return_scaled=False` with `dtype='int16'` when, under your version, `return_scaled=True` by default. That break I am against, since I know we already make use of that exact function call at several places in currently supported projects, so it would be quite the annoyance to go back in and correct them all. And if it's annoying for me, who knows how many other downstream users would have their workflows disrupted by it.

Comment (alejoe91) [reply spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py:66]: I agree that we should keep in mind backcompatibility, but note that `return_scaled` already breaks some stuff in other extractors (e.g. NWB write function now requires it..). I think that SI is still immature enough to properly make these changes that would make the API better as a whole!

**Comment by CodyCBakerPhD** [comment]
@alejoe91 

1) `BinDat` no longer has a `self._gain` attribute, and instead utilizes the channel properties for gains. Also added typing annotations. Also added docstring.

2) For `Intan`, `dtype` is removed from `__init__`, and always adds gain/offset information as channel properties. `dtype` is kept in `get_traces`, but `return_scaled` has precedence over it, up to an attempt to pass `return_scaled=False`, `dtype='float',` which throws an error. If `return_scaled=False` and `dtype` is unspecified it defaults to `uint16`.

3) Also reviewed and fixed the impact this argument has on `NwbRecordingExtractor.write_recording`, which was always fundamentally built to operate on unscaled traces, storing gain information separately from the trace call.

File: spikeextractors/extractors/bindatrecordingextractor/bindatrecordingextractor.py [hunk 3]
```diff
@@ -77,66 +135,96 @@ def get_traces(self, channel_ids=None, start_frame=None, end_frame=None, return_
             exp_idx = self._dtype.find('int') + 3
             exp = int(self._dtype[exp_idx:])
             recordings = recordings.astype('float32') - 2**(exp - 1)
-        if self._gain is not None and return_scaled:
-            recordings = recordings * self._gain
+        if 'gain' in self.get_shared_channel_property_names() and return_scaled:
+            recordings = recordings * np.reshape(self.get_channel_gains(channel_ids=channel_ids), (len(channel_ids), 1))
```
Comment (alejoe91) [inline spikeextractors/extractors/bindatrecordingextractor/bindatrecordingextractor.py:139]: @CodyCBakerPhD a nicer way is the following:
```suggestion
            recordings = recordings * self.get_channel_gains(channel_ids=channel_ids)[:, None]
```

File: spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py [hunk 4]
```diff
@@ -61,16 +77,70 @@ def get_sampling_frequency(self):
...
+            Any of 'float', 'int16', or 'uint16'. return_scaled must be False to return 'int16' or 'uint16'.
+        Returns
+        ----------
+        traces: numpy.ndarray
+            A 2D array that contains all of the traces from each channel.
+            Dimensions are: (num_channels x num_frames)
+        """
+        assert dtype in ['float', 'int16', 'uint16', None], "'dtype' must be either 'float', 'int16', or 'uint16'!"
+
+        if return_scaled:
```
Comment (alejoe91) [inline spikeextractors/extractors/intanrecordingextractor/intanrecordingextractor.py:127]: nice!

File: spikeextractors/extractors/nwbextractors/nwbextractors.py [hunk 5]
```diff
@@ -849,8 +847,7 @@ def add_all_to_nwbfile(recording: se.RecordingExtractor, nwbfile=None,
             nwbfile=nwbfile,
             use_timestamps=use_timestamps,
             metadata=metadata,
-            write_as_lfp=write_as_lfp,
-            return_scaled=return_scaled
```
Comment (alejoe91) [inline spikeextractors/extractors/nwbextractors/nwbextractors.py:853]: Why not leaving the option to use `return_scaled=True`? I agree it should be `False` by default though

**Comment by alejoe91** [comment]
Hey @CodyCBakerPhD 

Almost ready to go! I have a few minor comments :)

File: spikeextractors/extractors/nwbextractors/nwbextractors.py (hunk 5 above)
Comment (CodyCBakerPhD) [reply spikeextractors/extractors/nwbextractors/nwbextractors.py:853]: Because the trace access here is fundamentally intended to be False, not just by default. Nwb series objects always behave like the equivalent of `get_traces(..., return_scaled=False)` with `gains` set. The only issue we would encounter here is if a given extractor doesn't support `return_scaled=False` (say, it only records them as pre-converted floats), but if no extractor comes immediately to mind on that we can always wait until the situation arises. Even in that case, you'd simply write the series with `return_scaled=True` and conversion factor of `1e-6`.

File: spikeextractors/extractors/bindatrecordingextractor/bindatrecordingextractor.py (hunk 3 above)
Comment (CodyCBakerPhD) [reply spikeextractors/extractors/bindatrecordingextractor/bindatrecordingextractor.py:139]: Ahh, very nice!

**Comment by luiztauffer** [comment]
hi guys, is this any close to be merged?

**Comment by alejoe91** [comment]
@CodyCBakerPhD @luiztauffer Yes merging this to the `return_scaled` branch. Since this requires also the other packages to change, I believe it's going to take a bit more time before we merge to master.
